Modules/_posixsubprocess.c

```diff
@@ -677,6 +712,84 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
         cwd_obj2 = NULL;
     }
 
+    if (call_setgroups) {
+#ifdef HAVE_SETGROUPS
+        if (groups_list != Py_None) {
+            Py_ssize_t i;
+            unsigned long gid;
+
+            if (!PyList_Check(groups_list)) {
+                PyErr_SetString(PyExc_TypeError,
+                        "setgroups argument must be a list");
+                goto cleanup;
+            }
+            num_groups = PySequence_Size(groups_list);
+
+            if (num_groups < 0)
+                goto cleanup;
+
+            if (num_groups > MAX_GROUPS) {
+                PyErr_SetString(PyExc_ValueError, "too many groups");
+                goto cleanup;
+            }
+
+            if ((groups = PyMem_RawMalloc(num_groups * sizeof(gid_t))) == NULL) {
+                PyErr_SetString(PyExc_MemoryError,
+                        "failed to allocate memory for group list");
+                goto cleanup;
+            }
+
+            for (i = 0; i < num_groups; i++) {
+                PyObject *elem;
+                elem = PySequence_GetItem(groups_list, i);
+                if (!elem)
+                    goto cleanup;
+                if (!PyLong_Check(elem)) {
+                    PyErr_SetString(PyExc_TypeError,
+                                    "groups must be integers");
+                    Py_DECREF(elem);
+                    goto cleanup;
+                } else {
+                    /* In posixmodule.c UnsignedLong is used as a fallback value
+                    * if the value provided does not fit in a Long. Since we are
+                    * already doing the bounds checking on the Python side, we
+                    * can go directly to an UnsignedLong here. */
+                    gid = PyLong_AsUnsignedLong(elem);
+                    if ((gid == (unsigned long)-1 && PyErr_Occurred() != NULL)
+                            || (gid > INT_MAX)) {
```

Comment:
Hm, `INT_MAX` bound is actually too low: `uid_t` is unsigned 32-bit on modern systems, so `sudo python3 -c 'import os; os.setgroups([2**32 - 2])'` works on my Linux. (`2**32 - 1 == (uid_t)-1` doesn't work because it's a special value).

@gpshead Is there any reason we couldn't use `_Py_Gid_Converter`? It's used in `Modules/posixmodule.c` and `Modules/grpmodule.c`.

Reply:
Using `_Py_Gid_Converter` seems like a good idea.

Reply:
I migrated this to using `_Py_Gid_Converter`